### PR TITLE
API auxiliary resources

### PR DESCRIPTION
**Pull request by myabc**
https://www.openproject.org/work_packages/14053

### CODE REVIEW

File: lib/api/v3/root_representer.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+      link 'project' do
+        {
+          href: "#{root_url}v3/api/project/{project_id}",
+          templated: true
+        }
+      end
+
+      link 'statuses' do
+        "#{root_url}v3/api/statuses"
```
Comment (ulferts) [inline lib/api/v3/root_representer.rb]: all links in this representer have v3 and api switched.

Which brings up the question of whether we need something that will take care of:
- including OpenProject::StaticRouting::UrlHelpers
- the path fragment "#{root_url}api/v3"

File: lib/api/root.rb [hunk 2]
```diff
@@ -34,8 +34,16 @@
 module API
   class Root < Grape::API
     prefix :api
+
+    class Formatter
+      def call(object, env)
+        object.respond_to?(:to_json) ? object.to_json : MultiJson.dump(object)
+      end
+    end
+
```
Comment (mtakac2) [inline lib/api/root.rb:43]: :+1:

File: lib/api/v3/priorities/priority_collection_representer.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  module V3
+    module Priorities
+      class PriorityCollectionRepresenter < Roar::Decorator
+        include Roar::Representer::JSON::HAL
+        include OpenProject::StaticRouting::UrlHelpers
+
+        self.as_strategy = API::Utilities::CamelCasingStrategy.new
+
+        link :self do
+          "#{root_url}api/v3/priorities"
```
Comment (mtakac2) [inline lib/api/v3/priorities/priority_collection_representer.rb]: I think we've agreed on using just path in the links.

File: lib/api/v3/priorities/priorities_api.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+module API
+  module V3
+    module Priorities
+      class PrioritiesAPI < Grape::API
+
+        resources :priorities do
+          before do
+            @priorities = IssuePriority.all
+            @priorities.map! { |priority| PriorityModel.new(priority) }
+          end
```
Comment (mtakac2) [inline lib/api/v3/priorities/priorities_api.rb:39]: Are we going to have some other endpoints for priorities or just the index? If we are, we should probably move the code from the before block to the get endpoint itself.

File: lib/api/v3/priorities/priority_representer.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        include Roar::Representer::JSON::HAL
+        include Roar::Representer::Feature::Hypermedia
+        include OpenProject::StaticRouting::UrlHelpers
+
+        self.as_strategy = API::Utilities::CamelCasingStrategy.new
+
+        property :_type, exec_context: :decorator
+
+        property :id, getter: -> (*) { model.id }, render_nil: true
+        property :name
```
Comment (mtakac2) [inline lib/api/v3/priorities/priority_representer.rb:46]: I guess name cannot be nil, right? Anyway, wouldn't be safer to just use `render_nil: true` everywhere?

File: lib/api/v3/projects/project_model.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      class ProjectModel < Reform::Form
+        include Coercion
+
+        property :identifier,  type: String, virtual: true
+        property :name,        type: String
+        property :description, type: String
+        property :homepage,    type: String
+
+        property :created_on,  type: DateTime, virtual: true
+        property :updated_on,  type: DateTime, virtual: true
```
Comment (mtakac2) [inline lib/api/v3/projects/project_model.rb:45]: Should be aliased to `created_at` and `updated_at` to be consistent with other resources. Although I am not sure if it should be done in model on in representer.

File: lib/api/v3/projects/project_representer.rb [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        include Roar::Representer::Feature::Hypermedia
+        include OpenProject::StaticRouting::UrlHelpers
+
+        self.as_strategy = API::Utilities::CamelCasingStrategy.new
+
+        property :_type, exec_context: :decorator
+
+        link :self do
+          {
+            href: "#{root_url}api/v3/projects/#{represented.model.id}",
```
Comment (mtakac2) [inline lib/api/v3/projects/project_representer.rb]: Again - only path.

File: lib/api/v3/projects/project_representer.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            title: "#{represented.name}"
+          }
+        end
+
+        link 'versions' do
+          "#{root_url}api/v3/projects/#{represented.model.id}/versions"
```
Comment (mtakac2) [inline lib/api/v3/projects/project_representer.rb]: Path

**Comment by mtakac2** [comment]
@ulferts I like it. There are few really good things. Especially the formatter stuff. I have only few little comments, mainly about links (I think we should use just paths everywhere) and properties naming (inconsistent accross resources). But I think we can merge it as it is and maybe change it later? Should I go on and merge it?

**Pull request merged**